src/calc/src/expr_manager.rs

```diff
@@ -0,0 +1,540 @@
+                eval = if token.item.skip_conv {
+                    // Binary notation. If we skip convertion.
+                    format!("{}{}{}", operand1, token.item.eval, operand2)
+                } else {
+                    // Binary function notation.
+                    format!("{}({},{})", token.item.eval, operand1, operand2)
+                }
+            }
+            _ => {}
+        }
+
+        // Change the top of the stack to the new compound operation (insead
+        // of creating new one).
+        let top = eval_stack.last_mut().unwrap();
+        top.btn = Btn::Evaluate; // Mark as compound operation.
+        top.item.eval = eval; // Set the newly generated evaluate string. Note that we don't
+                              // care about disp string as it is not used in the convertion
+                              // process.
+        top.item.skip_conv = token.item.skip_conv;
+        top.item.priority = token.item.priority;
+        top.arity = token.arity;
+
+        Ok(())
+    }   // push_func()
+
+    /// Construct final evaluation string from the tokens in postfix order.
+    fn to_eval_str(&self, postfix: &Vec<&Token>) -> Result<String, String> {
+        if postfix.is_empty() {
+            // This should never happen, as it is already handled in `get_eval_str()`.
+            // But just to know, if something has gone wrong.
+            panic!("Postfix is empty.");
+        }
+
+        // Holds the evaluated compound tokens.
+        let mut eval_stack: Vec<Token> = Vec::new();
+
+        // Evaluate tokens. Push the non-operation tokens onto the stack
+        // and then convert them to operations based on the operation tokens.
+        // The resulting compound token will be left on the top.
+        for token in postfix {
+            match token.btn {
+                // Non-operation tokens. Just push them onto the stack.
+                Btn::Num(_) | Btn::Comma | Btn::Const(_) | Btn::Random => {
+                    eval_stack.push((*token).clone())
+                }
+                // Operation tokens. This will pop the non-operation tokens (number depends on `token.arity`)
+                // and create a compound token on the top of the stack.
+                Btn::UnaryOpt(_) | Btn::BinOpt(_) => self.push_func(&mut eval_stack, token)?,
+                // TODO: Handle ans
+                Btn::Ans => todo!(),
+                // There should only be operation and non-operation tokens on the stack.
+                _ => panic!("Invalid token on the evaluate stack. {:?}", token),
+            };
+        }
+
+        // Pop the resulting compound token and get its eval string.
+        // We can safely unwrap it, because the `postfix` is never empty.
+        Ok(eval_stack.pop().unwrap().item.eval)
+    }
+
+    /// Convert tokens to postfix notation.
+    fn to_postfix<'a>(&'a self, tokens: &'a Vec<Token>) -> Result<Vec<&'a Token>, String> {
```

Comment:
Use `&'a [Token]` instead of `&'a Vec<Token>`.
They are basically the same but `&'a [Token]` can be more generic (it still works if the `Vec<Token>` changes into other vector type like `VecDeque`) and it's the rusty way to do

Reply:
Got it. This is called Slice right? Also can you explain, why do I need to put a lifetime here? I thought that if parameters and return type have the same lifetime as `self`, then I don't need to specify it..

Reply:
Yes, if the lifetime in parameter and return type is the same, they are automatically have the same lifetime.
But here you have *multiple* parameters with lifetime, those parameters may have *different* lifetime and the compiler **cannot** decide which one to use for the return type, that's why you have to specify it.

Reply:
Oh, ok.